BUG: Update MSVC compiler version for IntelCCompiler

r-mike opened the pull request:
Build numpy with intelw compiler currently fails with the error:
Microsoft Visual C++ 14.1 is required. Get it with "Microsoft Visual C++ Build Tools": https://visualstudio.microsoft.com/downloads/
even when the VS toolset 14.1 is installed. The IntelCCompiler uses
an old MSVC version to compile. This commit updates the MSCV compiler version, which fixes the error.

mikeruf-philips committed:
Update MSVC compiler version for IntelCCompiler

Build numpy with intelw compiler currently fails with the error:
Microsoft Visual C++ 14.1 is required. Get it with "Microsoft Visual C++ Build Tools": https://visualstudio.microsoft.com/downloads/
even when the VS toolset 14.1 is installed. The IntelCCompiler uses
an old MSVC version to compile. This commit updates the MSCV compiler version, which fixes the error.

mattip commented:
Why do you need the new file? It seems exactly like `numpy/distutils/msvc9compiler.py`, or did I miss something subtle

mattip commented on numpy/distutils/intelccompiler.py:
```diff
@@ -92,9 +93,9 @@ def __init__(self, verbose=0, dry_run=0, force=0):
-            self.linker = self.find_exe('xilink')
+            self.cc = _find_exe('icl.exe')
+            self.lib = _find_exe('xilib')
+            self.linker = _find_exe('xilink')
```
At least for new code (python 3) `_find_exe` should be spelled [`shutil.which`](https://docs.python.org/3/library/shutil.html) to avoid using "private" interfaces. If this is to be backported, then [`distutils.spawn.find_executable function`] is still a better, although seemingly undocumented, alternative.

r-mike replied:
Since this is a backport-candidate I think I should use `distutils.spawn.find_executable`. There is also another `find_executable ` in `numpy.distutils.exec_command`. I compare results of the different `find_executable` on Monday.

r-mike replied:
With the new commit I use `distutils.spawn.find_executable`. Since `distutils._msvccompiler` returns the original program name if no path is found, I created a wrapper around `distutils.spawn.find_executable` to ensure that the both functions are having the same results. A few tests are failing at the moment, but I can't see how my changes are causing these tests to fail. Do I miss something?

r-mike commented:
Yes, it is almost the same file, except for the import `from distutils._msvccompiler import MSVCCompiler as _MSVCCompiler`. My first thought was to make a new file for keeping support with the older versions of VS like in python `distutils`. The header in `distutils._msvccompiler` says: 
```
"""distutils._msvccompiler

Contains MSVCCompiler, an implementation of the abstract CCompiler class
for Microsoft Visual Studio 2015.

The module is compatible with VS 2015 and later. You can find legacy support
for older versions in distutils.msvc9compiler and distutils.msvccompiler.
"""
```
However, I found a comment in `distutils._msvccompiler`, which says that VS 2005 and VS 2008 is also supported: `# ported to VS 2005 and VS 2008 by Christian Heimes`.  Since `distutils._msvccompiler` then should support >=VS2005 `numpy/distutils/msvc9compiler.py` should be obsolete and can be replaced by `numpy/distutils/_msvccompiler.py`. Any thoughts on this?

Commit:
Avoid using private interfaces from distutils._msvccompiler.

Implemented _find_exe with the same behaviour as distutils._msvccompiler._find_exe.

mattip commented on numpy/distutils/_msvccompiler.py:
```diff
@@ -0,0 +1,64 @@
+    def initialize(self, plat_name=None):
+        # The 'lib' and 'include' variables may be overwritten
+        # by MSVCCompiler.initialize, so save them for later merge.
+        environ_lib = os.getenv('lib')
```
What happens when this key is missing? Maybe `os.environ['libs']` -> `os.environ.get('libs', '')`, and below as well.

charris commented:
@r-mike I'm going to close this. If you still need this enhancement, feel free to open an new PR.
